ouroboros-consensus/src/Ouroboros/Consensus/Protocol/HardFork/OuroborosTag.hs

```diff
@@ -0,0 +1,178 @@
+              fmap BeforeFork <$>
+                checkIsLeader cfgP1 slotNo lviewP1 cstateP1
+
+        -- If it's time to switch, however, we should produce a leader proof for
+        -- protocol p2 rather than p1.
+          | otherwise -> runAfter $ do
+              let lviewP2  = ledgerViewAfterFork cfg lviewP1
+                  cstateP2 = chainStateAfterFork cfg cstateP1
+              fmap AfterFork <$>
+                checkIsLeader cfgP2 slotNo lviewP2 cstateP2
+
+        Nothing ->
+          error "checkIsLeader: inconsistent ledger view and chain state"
+
+  -- Apply header
+  --
+  -- Applying a header is the second place where we might detect that we have to
+  -- switch; in 'checkIsLeader' we must check if /we/ should be responsible for
+  -- initiating the hard fork, but here we might find that somebody else has.
+
+  applyChainState cfg@(HardForkCfg cfgP1 cfgP2) lview hdr cstate =
+      case (forkedPair lview (getForkedHeader cfg hdr), cstate) of
+        -- Before the fork. Delegate to P1
+        (Just (BeforeFork (lviewP1, hdrP1)), BeforeFork cstateP1) ->
+          mapExcept BeforeFork BeforeFork $
+            applyChainState cfgP1 lviewP1 hdrP1 cstateP1
+
+        -- After the fork. Delegate to P2
+        (Just (AfterFork (lviewP2, hdrP2)), AfterFork afcs) ->
+          mapExcept AfterFork AfterFork $
+            fmap (AFCS.dropSnapshotIfRedundant cfgP2 hdrP2) $
+              AFCS.update afcs $ applyChainState cfgP2 lviewP2 hdrP2
+
+        -- The ledger view is already updated, but the chain state hasn't yet
+        -- We assume that a "premature fork" would be detected by the ledger.
+        (Just (AfterFork (lviewP2, hdrP2)), BeforeFork cstateP1) -> do
+          let afcs = AFCS.init cfg hdrP2 cstateP1
+          mapExcept AfterFork AfterFork $
+            AFCS.update afcs $ applyChainState cfgP2 lviewP2 hdrP2
+
+        -- Error cases
+
+        (Just (BeforeFork _), AfterFork _) ->
+          -- We have a ledger view from /before/ the fork, but a chain state
+          -- from /after/ the fork. This can't happen: we only construct the
+          -- after-fork chain state when receiving an after-fork leger view.
+          -- TODO: Think through what happens during rollback.
+          error "applyChainState:  inconsistent ledger view and chain state"
+        (Nothing, _) ->
+          -- We have a before-fork header and an after-fork ledger view, or
+          -- an after-fork header and a before-fork ledger view. Either way,
+          -- this cannot happen, because we are getting the /updated/ ledger
+          -- view, and that update would have failed in these cases.
+          error "applyChainState: impossible updated ledger view"
+    where
+      mapExcept :: (e -> e') -> (a -> b) -> Except e a -> Except e' b
+      mapExcept f g = withExcept f . fmap g
+
+  -- Rewind chain state
+
+  rewindChainState (HardForkCfg cfgP1 cfgP2) cstate mSlotNo =
```

Comment:
This rewind window might be too large (P2 might restrict the window, but if we have a snapshot, we could rollback anyway). Think about whether we could tighten this (related: `dropSnapshotIfNeeded`).